certbot/plugins/dns_cloudflare.py

```diff
@@ -0,0 +1,323 @@
+
+        if self.email:
+            setattr(self.config, self.dest('email'), self.email)
+        else:
+            raise errors.PluginError('Cloudflare account email address required to proceed.')
+
+        configured_api_key = self.conf('api-key')
+        if configured_api_key:
+            self.api_key = configured_api_key
+        else:
+            self.api_key = self._prompt_for_data('API key')
+
+        if self.api_key:
+            setattr(self.config, self.dest('api-key'), self.api_key)
+        else:
+            raise errors.PluginError('Cloudflare account API key required to proceed.')
+
+    @staticmethod
+    def _prompt_for_data(label):
+        """
+        Prompt the user for a piece of information.
+
+        :param string label: The user-friendly label for this piece of information.
+        :returns: The user's response (guaranteed non-empty).
+        :rtype: string
+        """
+
+        display = zope.component.getUtility(interfaces.IDisplay)
+
+        while True:
+            code, response = display.input(
+                'Input Cloudflare account {0}'.format(label),
+                force_interactive=True)
+            if code == display_util.HELP:
+                # Displaying help is not currently implemented
+                return None
+            elif code == display_util.CANCEL:
+                return None
+            else:  # code == display_util.OK
+                if not response:
+                    display.notification('Please enter an {0}.'.format(label), pause=False)
+                else:
+                    return response
+
+    def _get_cloudflare_client(self):
+        """
+        Helper method to construct a `_CloudflareClient`.
+
+        Uses configured credentials.
+
+        :return: a new
+        :rtype: `_CloudflareClient`
+        """
+
+        return _CloudflareClient(self.email, self.api_key)
+
+
+class _CloudflareClient(object):
+    """
+    Encapsulates all communication with the Cloudflare API.
+    """
+
+    def __init__(self, email, api_key):
+        self.cf = CloudFlare.CloudFlare(email, api_key)
+
+    def add_txt_record(self, domain, record_name, record_content, record_ttl):
+        """
+        Add a TXT record using the supplied information.
+
+        :param string domain: The domain to use to look up the Cloudflare zone.
+        :param string record_name: The record name (typically beginning with '_acme-challenge.').
+        :param string record_content: The record content (typically the challenge validation).
+        :param int record_ttl: The record TTL (number of seconds that the record may be cached).
+        :raises: errors.PluginError if an error occurs communicating with the Cloudflare API
+        """
+
+        zone_id = self._find_zone_id(domain)
+
+        data = {'type': 'TXT',
+                'name': record_name,
+                'content': record_content,
+                'ttl': record_ttl}
```

Comment:
How close do you think Lexicon is to a state where we can use it for all/more of our DNS plugins? They're not packaged in any major distro either so we'd probably have to vendorize them as well or make the functionality only available to a smaller subset of Certbot users. With this in mind, if we could easily work with upstream or fork it to get the functionality we wanted, I think it may be worth investigating.

> If we address #1279, I think this would be worth considering regardless of the dependency discussion; it seems like few users will ever use all DNS plugins.

My instinct is to put it in the base Certbot package if there are a minimal number of dependencies added/vendorized. If we can add the functionality to the core package without significantly increasing bloat, it should be done to make it easy for users to get the functionality they need, otherwise, it should go into its own package. For example, we decided for route53 we don't want to add/vendorize `boto3` and will be going another route.

> logging is just a standard python library, but yes: we could vendorize cloudflare.

`logging` is but `cloudflare` depends on `requests`, `logger`, `pyyaml`, and `future` ([source](https://github.com/cloudflare/python-cloudflare/blob/master/setup.py#L35)). We already depend on `requests` but `logger`, `pyyaml`, and `future` would be new dependencies for us. `logger` isn't packaged anywhere so we'd probably have to vendorize it in addition to `cloudflare`. `pyyaml` is pretty thoroughly packaged, but it'd still be a new (vendorized) dependency. `future` is somewhere in the middle between `logger` and `pyyaml` about how widely a distro package is available.

I reached out to our Debian package maintainer to get his thoughts, but the need to add 4 new dependencies to support Cloudflare has me a little hesitant of taking this approach or putting it in the core Certbot package.

> One option that I don't think I like, but is worth considering: we could make cloudflare an optional dependency and hide/disable the dns-cloudflare plugin if it is not present.

Yeah that's definitely an option. Previously we've taken the approach of separate packages as you can see by `certbot-apache` and `certbot-nginx`, but we could change our approach here if we wanted.